sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/V1Writes.scala

```diff
@@ -98,16 +98,27 @@ object V1Writes extends Rule[LogicalPlan] with SQLConfHelper {
     assert(empty2NullPlan.output.length == query.output.length)
     val attrMap = AttributeMap(query.output.zip(empty2NullPlan.output))
 
-    // Rewrite the attribute references in the required ordering to use the new output.
-    val requiredOrdering = write.requiredOrdering.map(_.transform {
-      case a: Attribute => attrMap.getOrElse(a, a)
-    }.asInstanceOf[SortOrder])
+    // Rewrite the attribute references to use the new output.
+    def rewriteAttributes[T <: Expression](expr: T): T = {
+      expr.transform {
+        case a: Attribute => attrMap.getOrElse(a, a)
+      }.asInstanceOf[T]
+    }
+    val clusteringSpec = write.clusteringSpec.map { spec =>
+      ClusteringSpec(
+        clusteringKeys = spec.clusteringKeys.map(rewriteAttributes),
+        sortKeys = spec.sortKeys.map(rewriteAttributes)
+      )
+    }
+    val clusteringExpressions = clusteringSpec.map(_.expressions).getOrElse(Seq.empty)
```

Comment:
shouldn't we call `toSorts`? We should respect the sort order of the sorting columns.

Reply:
OK it was the previous behavior, but it seems wrong?

Reply:
I think `isOrderingMatched` should take `ClusteringSpec` and only ignore the `SortDirection` for clustering keys.

Reply:
The `expressions` function returns both expressions of `clusteringKeys` and `sortKeys`, we use it here because if the `outputOrdering` matched all expressions, then neither `Clustering` nor `Sort` node needs to be added here.

And `isOrderingMatched` is also called in `FileFormatWriter`, so I think it could still accept the `Seq[Expression]` parameter : 
https://github.com/apache/spark/blob/0123a5ecbe6d4075b0738e9d2faac354f2cbd008/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala#L156-L160

Reply:
I'm fine with it because it's the existing behavior, but I believe there is a bug here. See https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala#L389-L391

The table scan node assumes the bucket sort columns are always ascending, but here we may sort the data in descending order, because we may remove the `Clustering` node if the plan is already sorted by the sort columns descendingly.